plugin/storage/es/ilm/policymanager.go

```diff
@@ -0,0 +1,146 @@
+	"strconv"
+	"time"
+
+	"github.com/jaegertracing/jaeger/pkg/es"
+	"github.com/jaegertracing/jaeger/pkg/es/client"
+	"github.com/jaegertracing/jaeger/pkg/es/filter"
+)
+
+const (
+	DefaultIlmPolicy    = "jaeger-default-ilm-policy"
+	DefaultIsmPolicy    = "jaeger-default-ism-policy"
+	writeAliasSuffix    = "write"
+	readAliasSuffix     = "read"
+	rolloverIndexSuffix = "000001"
+	ilmVersionSupport   = 7
+	ilmPolicyFile       = "ilm-policy.json"
+	ismPolicyFile       = "ism-policy.json"
+)
+
+var ErrIlmNotSupported = errors.New("ILM is supported only for ES version 7+")
+
+// PolicyManager manages the environment for ILM or ISM policy
+type PolicyManager struct {
+	client                         func() es.Client
+	prefixedIndexNameWithSeparator string
+}
+
+// Init makes the jaeger ready for automatic rollover by using ILM by creating
+// initial rollover indices and read-write aliases
+func (p *PolicyManager) Init() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	err := p.init(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewPolicyManager creates the policy manager with appropriate version and prefixedIndexNameWithSeparator.
+// prefixedIndexNameWithSeparator is the prefix with separator. For example if index prefix is jaeger-main
+// and policy manager is called for span indices, then prefixedIndexNameWithSeparator will be jaeger-main-jaeger-span-
+func NewPolicyManager(cl func() es.Client, prefixedIndexNameWithSeparator string) *PolicyManager {
+	return &PolicyManager{
+		client:                         cl,
+		prefixedIndexNameWithSeparator: prefixedIndexNameWithSeparator,
+	}
+}
+
+func (p *PolicyManager) init(ctx context.Context) error {
+	index := p.initialRolloverIndex()
+	err := p.createIndexIfNotExists(ctx, index)
+	if err != nil {
+		return err
+	}
+	jaegerIndices, err := p.getJaegerIndices(ctx)
+	if err != nil {
+		return err
+	}
+	readAlias := p.readAliasName()
+	writeAlias := p.writeAliasName()
+	var aliases []client.Alias
+	if !filter.AliasExists(jaegerIndices, readAlias) {
+		aliases = append(aliases, client.Alias{
+			Index:        index,
+			Name:         readAlias,
+			IsWriteIndex: false,
+		})
+	}
+	if !filter.AliasExists(jaegerIndices, writeAlias) {
+		aliases = append(aliases, client.Alias{
+			Index:        index,
+			Name:         writeAlias,
+			IsWriteIndex: true,
+		})
+	}
+	for _, alias := range aliases {
+		_, err = p.client().CreateAlias(alias.Name).Index(alias.Index).IsWriteIndex(alias.IsWriteIndex).Do(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *PolicyManager) createIndexIfNotExists(ctx context.Context, index string) error {
+	exists, err := p.client().IndexExists(index).Do(ctx)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		_, err = p.client().CreateIndex(index).Do(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *PolicyManager) getJaegerIndices(ctx context.Context) ([]client.Index, error) {
```

Comment:
what is this function doing and why? I thought the whole point of ILM policy was that we don't have to worry about individual indices